src/optimizer/rule/LimitPushDownRule.cpp

```diff
@@ -0,0 +1,200 @@
+                                    const OptGroupExpr *groupExpr,
+                                    TransformResult *result) const {
+    auto pair = findMatchedGroupExpr(groupExpr);
+    auto projExpr = pair.second[0];
+    auto gnExpr = pair.second[1];
+
+    const auto limit = static_cast<const Limit *>(groupExpr->node());
+    const auto proj = static_cast<const Project *>(projExpr->node());
+    const auto gn = static_cast<const GetNeighbors *>(gnExpr->node());
+
+    int64_t limitRows = limit->offset() + limit->count();
+
+    auto newLimit = cloneLimit(qctx, limit);
+    auto newLimitExpr = OptGroupExpr::create(qctx, newLimit, groupExpr->group());
+
+    auto newProj = cloneProj(qctx, proj);
+    auto newProjGroup = OptGroup::create(qctx);
+    auto newProjExpr = newProjGroup->makeGroupExpr(qctx, newProj);
+
+    auto newGn = cloneGetNbrs(qctx, gn);
+    newGn->setLimit(limitRows);
+    auto newGnGroup = OptGroup::create(qctx);
+    auto newGnExpr = newGnGroup->makeGroupExpr(qctx, newGn);
+
+    newLimitExpr->dependsOn(newProjGroup);
+    newProjExpr->dependsOn(newGnGroup);
+    for (auto dep : gnExpr->dependencies()) {
+       newGnExpr->dependsOn(dep);
+    }
+
+    result->eraseAll = true;
+    result->eraseCurr = true;
+    result->newGroupExprs.emplace_back(newLimitExpr);
+    return Status::OK();
+}
+
+std::string LimitPushDownRule::toString() const {
+    return "LimitPushDownRule";
+}
+
+std::pair<bool, std::vector<const OptGroupExpr *>> LimitPushDownRule::findMatchedGroupExpr(
+    const OptGroupExpr *groupExpr) const {
+    std::vector<const OptGroupExpr *> matched;
+
+    auto node = groupExpr->node();
+    if (node->kind() != PlanNode::Kind::kLimit) {
+        return std::make_pair(false, matched);
+    }
+
+    const auto limit = static_cast<const Limit *>(node);
+    int64_t limitRows = limit->offset() + limit->count();
+
+    bool projFound = false;
+    const OptGroupExpr *projExpr = nullptr;
+    for (auto dep : groupExpr->dependencies()) {
+        for (auto expr : dep->groupExprs()) {
+            if (expr->node()->kind() == PlanNode::Kind::kProject) {
+                projFound = true;
+                projExpr = expr;
+                break;
+            }
+        }
+    }
+
+    if (!projFound) {
+        return std::make_pair(false, matched);
+    }
+
+    bool gnFound = false;
+    const OptGroupExpr *gnExpr = nullptr;
+    for (auto dep : projExpr->dependencies()) {
+        for (auto expr : dep->groupExprs()) {
+            if (expr->node()->kind() == PlanNode::Kind::kGetNeighbors) {
+                gnFound = true;
+                gnExpr = expr;
+                break;
+            }
+        }
+    }
+
+    if (!gnFound) {
+        return std::make_pair(false, matched);
+    }
+
+    const auto gn = static_cast<const GetNeighbors *>(gnExpr->node());
+    if (limitRows >= gn->limit()) {
+        return std::make_pair(false, matched);
+    }
+    matched.push_back(projExpr);
+    matched.push_back(gnExpr);
```

Comment:
Returned OptGroupExprs only contains the root GroupExpr rather than all touched GroupExprs

Reply:
Why? If just return projExpr, function LimitPushDownRule::transform will execute the following code again:

>     bool gnFound = false;
    const OptGroupExpr *gnExpr = nullptr;
    for (auto dep : projExpr->dependencies()) {
        for (auto expr : dep->groupExprs()) {
            if (expr->node()->kind() == PlanNode::Kind::kGetNeighbors) {
                gnFound = true;
                gnExpr = expr;
                break;
            }
        }
    }

Reply:
make sense. I'm thinking about whether need to collect these OptGroupExprs by patten match.